File: Problems/No.28_Implement_strStr()/implement_strstr().py
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        
        # store the length
        len_haystack = len(haystack)
        len_needle = len(needle)
        
        # preprocess
        if (len_haystack < len_needle):
            return -1
        
        # start compute
        for i in range(len_haystack - len_needle + 1):
            if (haystack[i:i+len_needle] == needle):
                return i
        
        return -1

File: Problems/No.28_Implement_strStr()/implement_strstr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        
        // store the current index of haystack and needle
        int h_index = 0;
        int n_index = 0;
        
        // start compute
        while (h_index < haystack.length()) {
            
            // if same then compare the next index else h_index++, n_index <- 0
            if (haystack[h_index] == needle[n_index]) {
                
                h_index++;
                n_index++;
                
            }
            else {
                
                h_index = h_index - n_index + 1;
                n_index = 0;
                
            }
            
            if (n_index == needle.length()) return h_index-n_index;
            
        }
        
        return -1;
        
    }
};
